b2sdk/utils/curl.py

```diff
@@ -0,0 +1,431 @@
+
+    def is_empty(self) -> bool:
+        with self.lock:
+            return len(self.deque) == 0
+
+    def write(self, buffer) -> int:
+        with self.lock:
+            if len(buffer) == 0:
+                return 0
+
+            # Whole buffer is stored as deque is implemented as double-linked list.
+            # This way we have fewer, (hopefully) larger elements.
+            self.deque.append(buffer)
+
+            # If we haven't informed everyone yet, it's the time. We've received
+            # first buffer of data. This is used to indicate that all the headers
+            # are already downloaded.
+            if not self.first_write_event.is_set():
+                self.first_write_event.set()
+
+            return len(buffer)
+
+    def read(self, size: Optional[int] = None) -> bytes:
+        with self.lock:
+            result = bytearray()
+
+            if len(self.deque) == 0:
+                return result
+
+            while True:
+                entry = self.deque.popleft()
+                new_len_result = len(result) + len(entry)
+
+                if new_len_result > size:
+                    oversize = new_len_result - size
+                    # Return over-sized buffer to the queue.
+                    oversize_buffer = entry[-oversize:]
+                    self.deque.appendleft(oversize_buffer)
+                    # Cut down next entry.
+                    entry = entry[:-oversize]
+
+                result += entry
+
+                if (size != -1 and len(result) == size) or len(self.deque) == 0:
+                    break
+
+            return result
+
+
+class HeaderReader:
+    # RFC 2086 Section 4.2 states:
+    # HTTP header fields (...) follow the same generic format as that given in Section 3.1 of RFC 822
+    #
+    # email.parser is made to parse RFC 822 headers. Only difference is in encodings. By default,
+    # email.parser uses ASCII encoding while RFC 2086 requires ISO-8859-1. We'll conform to the latter.
+    PARSER = email.parser.HeaderParser()
+    ENCODING = 'iso-8859-1'
+
+    def __init__(self):
+        self.lines = []
+        self.headers: Optional[CaseInsensitiveDict] = None
+
+    def add_header_line(self, line: bytes) -> None:
+        # First line of the header contains http version and status code.
+        # Last line of the header contains just `\r\n`. Removing both.
+        if line.startswith(b'HTTP/') or len(line.strip()) == 0:
+            return
+        self.lines.append(line)
+
+    def parse_headers(self) -> None:
+        if self.headers is not None:
+            return
+
+        parsed_header = self.PARSER.parsestr(
+            b''.join(self.lines).decode(self.ENCODING),
+            headersonly=True,
+        )
+        self.headers = CaseInsensitiveDict(parsed_header.items())
+
+
+@dataclass
+class CurlStreamer:
+    """
+    Simple Curl object wrapper.
+
+    Manages downloading on demand of headers and content. Can also be used in blocking mode.
```

Comment:
I understand neither `Manages downloading on demand of headers and content.` nor `Can also be used in blocking mode.`. Can you please rephrase or provide more explanation?

Reply:
I should probably add a nice, big block of explanation at the very start. Will do.